c7n/credentials.py

```diff
@@ -19,6 +20,37 @@
     'C7N_USE_STS_REGIONAL', '').lower() in ('yes', 'true')
 
 
+class CustodianSession(Session):
+
+    # track clients and return extant ones if present
+    _clients = {}
```

Comment:
Hmm does having this at the class level risk having some pollution of clients across role assumes?

Reply:
yeah, I've been wondering that, I think we need to key of a credential hash as well. we do resets between policy executions, which is out standard for member account role assumes etc, but for outputs we'll have some mix of clients across account boundaries in some cases.

Reply:
added credential hash to the cache key